### PR TITLE
e2e matmul testing for compilation infos on gpu

### PR DESCRIPTION
**Pull request by KoolJBlack**
Integrates e2e matmul testing for GPUs; intended to test the compilation verifiers for `LLVMGPUMatmulSimt` and `LLVMGPUMatmulTensorCore` pass pipelines. 

Changes mainly to `generate_e2e_matmul_tests.py`: In order to support GPU testing apart from CPU:
* `ShapesId::GPU_LARGE` option to `--shapes` for gpu shapes
* `--compilation_info` allows to specify the compiling pass pipeline for generated test cases and compilation infos

GPU tests make use of additional compilation info attrs added during generation. These are committed with other tests (`--compilation_info=none`)

### CODE REVIEW

**Comment by KoolJBlack** [comment]
@GMNGeoffrey the bazel build keeps triggering here for the cuda test even with the flags you mentioned. No idea how to restrict it.

File: iree/test/e2e/regression/BUILD
Comment (ScottTodd) [inline iree/test/e2e/regression/BUILD:247]: These are only added to the test, not the test deps.

https://github.com/google/iree/blob/e10830fb3a7dba1d77fe517b9620a0a1a797f7e3/build_tools/bazel/iree_trace_runner_test.bzl#L39-L87

Comment (ScottTodd) [reply iree/test/e2e/regression/BUILD:247]: Here, you want to tag the `iree_bytecode_module` as `manual` + `nokokoro` (or the cuda tags), as that requires the compiler.

Here is how the `iree_check_test` function skips cuda: https://github.com/google/iree/blob/e10830fb3a7dba1d77fe517b9620a0a1a797f7e3/build_tools/bazel/iree_check_test.bzl#L128-L132

Comment (ScottTodd) [reply iree/test/e2e/regression/BUILD:247]: Another option is to just omit this particular `iree_generated_trace_runner_test` from the BUILD file and put it in the `BAZEL_TO_CMAKE_PRESERVES_ALL_CONTENT_BELOW_THIS_LINE` section of the CMake file. (That will be cleaner once the `iree_generated_trace_runner_test` test are in their own dir, instead of next to the other e2e/regression/ tests)

Comment (KoolJBlack) [reply iree/test/e2e/regression/BUILD:247]: I've updated the `iree_trace_runner_test.bzl` to skip tests the use the cuda backend, since bazel doesn't support cuda.

**Comment by ThomasRaoux** [comment]
There are some failures on Turing build, it looks like this is compiling through TensorCore pipeline with a sm lower than 80. TensorCore pipeline requires "--iree-cuda-llvm-target-arch=sm_80" option and cannot run on Turing so we should disable the run until we upgrade to Ampere.

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [inline iree/test/e2e/regression/generate_e2e_matmul_tests.py:176]: Are you sure that you want to test only one shape where `M==K==N`?  I can understand that you want to test only one shape, and that you want it fairly large like this, but wouldn't you obtain a little more interesting test coverage by exercising `M`, `K`, `N` being distinct values, e.g. in case some compiler code inadvertently mixed up dims? @ThomasRaoux ?

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [inline iree/test/e2e/regression/generate_e2e_matmul_tests.py:242]: No dynamic-shapes testing at all for GPU at the moment? @ThomasRaoux ?

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [inline iree/test/e2e/regression/generate_e2e_matmul_tests.py:595]: Wouldn't it be more usual to have the default value be the empty string `""` , rather than a custom keyword `"none"` ? That would also allow writing more idiomatic code in my suggestion below, `if not compilation_info_id:`.

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [inline iree/test/e2e/regression/generate_e2e_matmul_tests.py:682]: An error should be generated if `compilation_info_id` is not one of the expected values. Otherwise, my undertanding is that at the moment, if one makes a typo in the `compilation_info` flag, that silently falls through to the below `generate_default` path.  If you accept my above suggestion to have the default value be `""` instead of `"none"` then that would be a matter of checking `if compilation_info_id:` .

**Review by bjacob — CHANGES_REQUESTED** [review]

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (ThomasRaoux) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:176]: I agree with Benoit, it is fine to start with a single shape but picking `M==K==N` reduces the coverage area for no good reason.

File: iree/test/e2e/regression/BUILD
Comment (KoolJBlack) [reply iree/test/e2e/regression/BUILD:135]: I marked the argument as required, so even though it has a default you still have to actually list it (or it throws an error). I agree this is kind of verbose, I can just make it optional and remove it elsewhere.

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (KoolJBlack) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:595]: Done. however note that the compilation_info_id is an enum so I must check for an explicit value

File: iree/test/e2e/regression/BUILD
Comment (KoolJBlack) [reply iree/test/e2e/regression/BUILD:234]: I made the "large" shapes a lot less large and these test should run substantially faster now.

Comment (bjacob) [reply iree/test/e2e/regression/BUILD:231]: FWIW the existing e2e matmul tests for CPU are a lot more expensive (test many more shapes) so if this is the concern, we can find more opportunity for improvement on that side.

File: iree/test/e2e/regression/CMakeLists.txt
Comment (bjacob) [inline iree/test/e2e/regression/CMakeLists.txt:327]: The `"driver=cuda"` label should already be generated based on the `DRIVERS` value.
https://sourcegraph.com/github.com/google/iree/-/blob/build_tools/cmake/iree_trace_runner_test.cmake?L23-24

File: iree/test/e2e/regression/BUILD
Comment (bjacob) [reply iree/test/e2e/regression/BUILD:244]: +1 though that could be fair to defer to a subsequent PR, particularly given the likelihood that that'll run into something new in CI ;-)

File: iree/test/e2e/regression/BUILD
Comment (bjacob) [reply iree/test/e2e/regression/BUILD:231]: Actually i had been meaning to do that , so i'll do it now - will report back here.
Anyway, it can't get cheaper than what this PR does for GPU -- just one shape, and not even particularly large either.

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:242]: ok

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (bjacob) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: Actually this PR is pretty good in this respect! The CompilationInfo/ileWorkgroupSizePair info is explicitly mangled into the test function's name, see `generate_function_name`.  That is logged by the test runner. So, though it will require reading the logs, it will be clear what function failed, and it will be easy for someone familiar with this test suite, and feasible for someone without prior familiarity, to trace it back to the specific TileWorkgroupSizePair.

File: iree/test/e2e/regression/BUILD
Comment (bjacob) [inline iree/test/e2e/regression/BUILD:180]: Wait, now that the new `NONE` enum value corresponds to the arg string value `""`,  which is the default value, don't you need to drop this flag here?  As it currently stands, iiuc, this `"none"` value is not one of the legal values so it should hit the `return ValueError` path that you just added?

**Review by bjacob — CHANGES_REQUESTED** [review]
I'm almost done nitpicking, thanks!

File: iree/test/e2e/regression/BUILD
Comment (bjacob) [reply iree/test/e2e/regression/BUILD:180]: Oh hey this is exactly what the CI is complaining about :-P
```
generate_e2e_matmul_tests.py: error: argument --compilation_info: invalid choice: 'none' (choose from '', 'LLVMGPUMatmulSimt', 'LLVMGPUMatmulTensorCore')
```

File: iree/test/e2e/regression/BUILD
Comment (bjacob) [reply iree/test/e2e/regression/BUILD:231]: https://github.com/google/iree/pull/8448

File: iree/test/e2e/regression/BUILD
Comment (KoolJBlack) [reply iree/test/e2e/regression/BUILD:180]: I missed this. Deleted the flag.

**Review by bjacob — APPROVED** [review]

File: iree/test/e2e/regression/BUILD
Comment (KoolJBlack) [reply iree/test/e2e/regression/BUILD:242]: Agreed, I'd like to enable Vulkan separately. So fare I've been verifying these verifiers with CUDA.

File: iree/test/e2e/regression/generate_e2e_matmul_tests.py
Comment (ScottTodd) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: > the generated CMakeLists would become enormous

If there are actually lots of test cases, that's okay. See for example this recently refactored test suite:
https://github.com/google/iree/tree/main/integrations/tensorflow/test/iree_tf_tests/math

The complexity I'm weary of here is using another python script to generate test cases, particularly now that this python script is 700 lines long. We already have multiple ways of defining and running tests (googletest, lit, check, etc.) split across two build systems, and this adds yet another. The TF integrations tests were using a cartesian product macro and it took quite a dedicated effort to move those over to the https://github.com/google/iree/blob/main/build_tools/testing/gen_test_matrix.py generator (https://github.com/google/iree/pull/6126) then to the https://github.com/google/iree/tree/main/integrations/tensorflow/test/python generator (https://github.com/google/iree/pull/8078). As we have a few of these, I'd like to see them look more alike and use the same underlying test runner (e.g. lit or check, depending on test configuration requirements).

Comment (ScottTodd) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: I don't want to stall this work too much, but we do have several efforts going on right now to simplify our infrastructure (mainly disconnecting compiler tests from the build phase and migrating from Kokoro to Buildkite), and I see signs here of us digging ourselves into another complexity hole.  I'd have to look more at the specifics of what these tests are trying to do to offer concrete suggestions, but I have a feeling that a combination of these could help:

* Move tests out of e2e/regression. They should not have been added there in the first place.
* Organize generator scripts across different test suites consistently
  * some combination of moving them to predictable places in the project structure and making them look and behave the same way
* Be okay with directories full of generated test files or verbose CMake files. Explicit tests are easy to understand at a glance
* Drop bazel support unless there is a really good reason to keep it

Comment (bjacob) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: good stuff! just to be clear, the things you point out here are preexisting, so this shouldn't block this PR now, right? The current state of these e2e matmul tests, including its location in `e2e/regression`, was introduced in PR #7347, reviewed by @benvanik and @GMNGeoffrey . We may need a meeting with all of us at the same time, to avoid continuing a telephone game.

Comment (ScottTodd) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: Most of it is preexisting, but this PR does increase iree/test/e2e/regression/generate_e2e_matmul_tests.py from 500 lines to 700 lines. That's the "step in the wrong direction" I'm worried about. There have been a few of these where I'm concerned about tech debt piling up. None of it feels like it should block any individual PR, but it all adds up and must be paid down at some point.

Comment (bjacob) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: Can you either file an issue or get us all in a meeting to record/discuss this?

Comment (KoolJBlack) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: I think this discussion is beyond the scope of this PR. This PR only ads 14 tests to a set of existing tests (literally multiple hundreds) using existing build paths. In terms of overall scope, the complexity added here is minor compared to what already exists. 

Massive restructuring of tests should be handled separately.

Comment (ScottTodd) [reply iree/test/e2e/regression/generate_e2e_matmul_tests.py:264]: The new `CompilationInfo` argument and associated plumbing through this already complex test generation script is within the scope of this PR. It is adding more complexity and indirection to a component that is already too complex. That's going to take more work to refactor and clean up later and will increase the cost of test maintenance.

This change on its own seems okay to me now, but I'll keep asking for this e2e matmul testing to be simplified until that happens.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**
